### DISCUSSION TITLE
Text with attachment (FileMessage & ImageMessage)

### DISCUSSION THREAD

**Discussion by gtu-myowin** [flyerhq/flutter_chat_ui#651] [Ideas & Requests]
Hello, it's just me again! I found out that the attachment message types (i.e, `FileMessage` and `ImageMessage`) is lacking a text field. The use case is

My web app has a messaging module and it allows users to type message after file is attached. 

<img width="866" alt="Screen Shot 2022-08-01 at 1 09 48 PM" src="https://user-images.githubusercontent.com/2696343/182089585-5c3d54d2-405e-44f8-b29d-3336d8b5b4c2.png">

And it displays like this on the web

<img width="787" alt="Screen Shot 2022-08-01 at 1 11 35 PM" src="https://user-images.githubusercontent.com/2696343/182089708-1c035c5b-1305-4f96-810c-8b82b0221f42.png">

But when I display that message on mobile app. I can only display just image (or just text). Yes, you do provide `customMessageBuilder` option to build custom message but it feels like reinventing the wheel if I build custom message to display both text and attachment (image & file).

![Simulator Screen Shot - iPhone 13 Pro Max - 2022-08-01 at 13 14 29](https://user-images.githubusercontent.com/2696343/182091985-b7c6aeb4-951a-43c3-be98-a31e7fc78acb.png)

**Comment by gtu-myowin** [comment]
I would like to elaborate my feature request more. 

I am fine with sending text & attachment separately from mobile. 

I just wanted to display both text & attachment. The feature request is not for sending but for displaying.

Thank you!

**Comment by demchenkoalex** [comment]
Hey, this wasn't planned initially because we try to do Telegram/WhatsApp/Signal messenger level, and they have full size previews with an option to add captions, which requires additional screen, which is not something we can get anywhere and we need to build it ourselves. We did what was easiest - attach a file/image and then write something and this is what Messenger does. I personally not a fan of the UIs you showed above, but I asked our designer and she says this is fine, after she does some research and proper design for it. So this might be added.

**Comment by gtu-myowin** [comment]
Thank you for your quick reply. Just to be clear, 

1. I don't mean to change the UI
2. It doesn't need to display as image caption on full size image view
3. I just wanted to display both text & image (or file) if user sent text & image/file at the same time.

I actually went ahead to implement this with `customMessageBuilder` to simply return a `Column()` widget of `TextMessage` and `ImageMessage` / `FileMessage` but then I realized that there is not reuse-able widget for those message types.

If I continue, I will also have to do tapping those message (download animation, preview, ...) which you have already done. 

Therefore, I thought to request for this feature.

**Comment by gtu-myowin** [comment]
I do really appreciate you and your team's hard work. I did my research before implementing chat module to my app and `flutter_chat_ui` is the coolest package I found out there. I don't consider what you and your team achieved so far easy at all.

**Comment by zamirszn** [comment]
> Thank you for your quick reply. Just to be clear,
> 
> 1. I don't mean to change the UI
> 2. It doesn't need to display as image caption on full size image view
> 3. I just wanted to display both text & image (or file) if user sent text & image/file at the same time.
> 
> I actually went ahead to implement this with `customMessageBuilder` to simply return a `Column()` widget of `TextMessage` and `ImageMessage` / `FileMessage` but then I realized that there is not reuse-able widget for those message types.
> 
> If I continue, I will also have to do tapping those message (download animation, preview, ...) which you have already done.
> 
> Therefore, I thought to request for this feature.

hey hello , how did you finally resolve this

**Comment by gtu-myowin** [comment]
@zamirszn I don't remember but I am quite sure that I have shown the image and message separately.

**Comment by zamirszn** [comment]
Any piece of code will help please

**Comment by zamirszn** [comment]
incase
![Screenshot (262)](https://github.com/flyerhq/flutter_chat_ui/assets/31168776/16fc662d-ca02-4863-bee7-615f9a7780be)
 anyone is looking to achieve this ,i found a working solution ,just pass any info you want use the metadata field like so
```
final message = types.ImageMessage(
        metadata: {"caption": chatMessage},
        author: types.User(
          id: senderId.toString(),
        ))
```

then in the Chat Widget use the imageMessageBuilder 

```
Chat(
                imageMessageBuilder: (
                  imageMessage, {
                  int? messageWidth,
                }) {
                  return CustomChatImageMessageWidget(
                    caption: imageMessage.metadata?["caption"],
                    imageMessage: imageMessage,
                    imageWidth: messageWidth,
                    imageUrl: imageMessage.uri,
                  );
                }}
```

**Comment by ykdromar** [comment]
@demchenkoalex  Any update on this?

**Comment by demchenkoalex** [comment]
if there will be an update you will see it :) I do not work on current implementation anymore, so all new features will go to the next major release, which is in the works
